data/builder/dirshard.go

```diff
@@ -0,0 +1,200 @@
+package builder
+
+import (
+	"fmt"
+	"hash"
+
+	bitfield "github.com/ipfs/go-bitfield"
+	"github.com/ipfs/go-unixfsnode/data"
+	"github.com/ipfs/go-unixfsnode/hamt"
+	dagpb "github.com/ipld/go-codec-dagpb"
+	"github.com/ipld/go-ipld-prime"
+	"github.com/multiformats/go-multihash"
+	"github.com/spaolacci/murmur3"
+)
+
+type shard struct {
+	// metadata about the shard
+	hasher  uint64
+	size    int
+	sizeLg2 int
+	width   int
+	depth   int
+
+	children map[int]entry
+}
+
+// a shard entry is either another shard, or a direct link.
+type entry struct {
+	*shard
+	*hamtLink
+}
+
+// a hamtLink is a member of the hamt - the file/directory pointed to, but
+// stored with it's hashed key used for addressing.
+type hamtLink struct {
+	hash hashBits
+	dagpb.PBLink
+}
+
+// BuildUnixFSShardedDirectory will build a hamt of unixfs hamt shards encoing a directory with more entries
+// than is typically allowed to fit in a standard IPFS single-block unixFS directory.
+func BuildUnixFSShardedDirectory(size int, hasher uint64, entries []dagpb.PBLink, ls *ipld.LinkSystem) (ipld.Link, error) {
+	// hash the entries
+	var h hash.Hash
+	var err error
+	// TODO: use the multihash registry once murmur3 behavior is encoded there.
+	// https://github.com/multiformats/go-multihash/pull/150
+	if hasher == hamt.HashMurmur3 {
+		h = murmur3.New64()
+	} else {
+		h, err = multihash.GetHasher(hasher)
+		if err != nil {
+			return nil, err
+		}
+	}
+	hamtEntries := make([]hamtLink, 0, len(entries))
+	for _, e := range entries {
+		name := e.Name.Must().String()
+		sum := h.Sum([]byte(name))
+		hamtEntries = append(hamtEntries, hamtLink{
+			sum,
+			e,
+		})
+	}
+
+	sizeLg2, err := logtwo(size)
+	if err != nil {
+		return nil, err
+	}
+
+	sharder := shard{
+		hasher:  hasher,
+		size:    size,
+		sizeLg2: sizeLg2,
+		width:   len(fmt.Sprintf("%X", size-1)),
```

Comment:
Hm, using a Sprintf just to take its length strikes me as odd. Consider `hex.EncodedLen`?

Reply:
how would you use `hex.EncodedLen` here? getting the length of `size-1` in bytes doesn't seem easier unless i'm missing something.

Reply:
Ah, you're right.